Clean up ROS + Gazebo Compatibility (#352)
gazebosim/docs: ros_installation.md

# Installing Gazebo with ROS

This document provides guidance on using different versions

> of ROS in combination with different versions of Gazebo. We recommended to read

Reviewer: ```suggestion
of ROS in combination with different versions of Gazebo. We recommend reading
```

it before installing [ros_gz](https://github.com/gazebosim/ros_gz).

# Picking the "Correct" Versions of ROS & Gazebo

If this is your first time using ROS and Gazebo, and you are not following

> specific instructions or a tutorials, we recommend using the latest long term

Reviewer: ```suggestion
specific instructions or tutorials, we recommend using the latest long term
```

support (LTS) versions of both ROS and Gazebo for the best user
experience. Successive versions of ROS and Gazebo releases are named
alphabetically, and the documentation for each version of ROS and Gazebo will
indicate if it is an LTS version. It is worth noting that each version of ROS
works best with a specific version of each Tier 1 platform. Tier one platforms
are platforms / host operating systems that are used in the development of ROS
and Gazebo. All of this information is outlined in
[REP-2000](https://www.ros.org/reps/rep-2000.html). 

To summarize, the best user experience is to use the latest LTS version of ROS
and the Tier 1 platform / operating system recommend for that version of ROS. If

> you host operating system does not match the Tier 1 operating system, consider

Reviewer: ```suggestion
your host operating system does not match the Tier 1 operating system, consider
```

# Installing Gazebo with ROS

This document provides guidance on using different versions
of ROS in combination with different versions of Gazebo. We recommended to read
it before installing [ros_gz](https://github.com/gazebosim/ros_gz).

# Picking the "Correct" Versions of ROS & Gazebo

If this is your first time using ROS and Gazebo, and you are not following
specific instructions or a tutorials, we recommend using the latest long term
support (LTS) versions of both ROS and Gazebo for the best user
experience. Successive versions of ROS and Gazebo releases are named
alphabetically, and the documentation for each version of ROS and Gazebo will
indicate if it is an LTS version. It is worth noting that each version of ROS
works best with a specific version of each Tier 1 platform. Tier one platforms
are platforms / host operating systems that are used in the development of ROS

> and Gazebo.All of this information is outlined in

Reviewer 2: ```suggestion
and Gazebo. All of this information is outlined in
```

[REP-2000](https://www.ros.org/reps/rep-2000.html). To summarize, the best user
experience is to use the latest LTS version of ROS and the Tier 1 platform /

> operating system recommend for that version of ROS. If you host operating system

Reviewer 2: Nitpick: this is a long paragraph that could be a bit hard to start the document with. Can we split it in a different one here:
```
operating system recommend for that version of ROS.

If you host operating system
```

Author: Fixed.

does not match the Tier 1 operating system, consider using the Tier 1 platform
in a virtual machine. This approach has the added benefit of not modifying your
host OS and allowing you to roll back your mistakes.

> At the time of writing, our recommendation is that new users install:

Reviewer 2: I've been trying not to suggest to the users a given platform (in this case Jammy) but instead suggest the best option given their platform. Here we are suggesting a platform for the first time. Not that I'm against it if you think that it helps new users.

Author: I think we should be transparent about what's easy to do and what's not easy to do. We get a lot of users who think they can install any version of ROS / Gazebo on any OS, or that installing ROS / Gazebo on Windows or Mac will just work out of the box. We also get a lot of users who are on a non-LTS Ubuntu install who think everything will just magically work, which isn't the case. Debian is the exception where things largely work except when they don't. I think we need to be really, really clear about what host OS will yield the best results out of the box.

* [Ubuntu Jammy 22.04](https://www.releases.ubuntu.com/jammy/)
* [ROS 2 Humble Hawksbill](https://www.ros.org/reps/rep-2000.html#humble-hawksbill-may-2022-may-2027)
* [Gazebo Fortress](https://gazebosim.org/docs/fortress/install)

# Summary of Compatible ROS and Gazebo Combinations

This table includes all currently supported versions of ROS and Gazebo. All
other ROS and Gazebo releases are end of life and we do not recommend their
continued use.

|                        	| **GZ Citadel (LTS)** 	| **GZ Fortress (LTS)** 	| **GZ Garden** 	|
|------------------------	|----------------------	|-----------------------	|---------------	|
| **ROS 2 Rolling**      	| X                    	| ✔️                     	| P             	|
| **ROS 2 Humble (LTS)** 	| X                    	| ✔️                     	| P             	|
| **ROS 2 Foxy (LTS)**   	| ✔️                    	| X                     	| X             	|
| **ROS 1 Noetic (LTS)** 	| ✔️                    	| P                     	| X             	|

* ✔️ - Recommended combination

> * X - Not recommended

Reviewer 2: ```suggestion
* X - Code incompatible
```
I think that we can go here with something stronger that "Not recommended" (which is fine just below in the P option). The X marks the options that we know they are not going to even compile since the code does not expect the given APIs.

Author: Maybe we say both. The goal is to make it clear that this is bad and dragons be here.

Author: fixed.

* P - Possible,*but not recommended*. These combinations of ROS and Gazebo can
  be made to work together, but some effort is required.

## Installing Gazebo with Older Versions of ROS

If you are planning on using a specific version of ROS and don't have a reason
to use a specific version of Gazebo, we recommend using the default version of
Gazebo available from the ROS repository when installing `ros_gz`. The following
command will install the correct version of Gazebo and `ros_gz` for your ROS
installation on a Linux system. You should replace `${ROS_DISTRO}` with your ROS
distribution (e.g. `humble`, `rolling`,`foxy`, etc).

```bash
sudo apt-get install ros-${ROS_DISTRO}-ros-gz
```

This command will install the Gazebo libraries recommended for your ROS
installation.

> **The rest of this document provides more information and other, alternative,

Reviewer 2: ```suggestion
<div class="warning">
The rest of this document provides more information and other, alternative,
```

Reviewer 2: The div warning displays a nice yellow block, see Note in the top of https://gazebosim.org/docs

installations pathways. We do not recommend the following approaches for new

> users.**

Reviewer 2: ```suggestion
users.
</div>
```

version of ROS or Gazebo.** However, if you must this version of ROS, it officially supports
Gazebo Edifice. To install Gazebo Edifice simply install `ros-galactic-ros-gz`
from `packages.ros.org`. This package will automatically install both Gazebo
Edifice and the ros-gz bridge.

### ROS 2 Foxy
ROS 2 Foxy officially supports Gazebo Citadel, meaning the recommended version
of Gazebo for ROS 2 Foxy is Citadel.To install Gazebo Citadel with Foxy simply
install `ros-foxy-ros-gz` from `packages.ros.org`. This package will
automatically install both Gazebo Citadel and the ros-gz bridge.

### ROS 1 Noetic
ROS Noetic officially supports Gazebo Citadel, meaning the recommended version
of Gazebo for ROS Noetic is Citadel.To install Gazebo Citadel with Noetic simply
install `ros-noetic-ros-gz` from `packages.ros.org`. This package will
automatically install both Gazebo Citadel and the ros-gz bridge.

## Using a specific and unsupported Gazebo version with ROS 2
<div class="warning">
<strong>Warning:</strong> Only use this approach if you absolutely need to run
a version of Gazebo that is not officially supported by your ROS distro. Using

> this approach will make it impossible to use the official ROS Ubuntu packages.

Reviewer 2: It is not that someone is not going to be able to use most of the ROS Ubuntu packages, but only a few subset. those ones that are using Gazebo in their building process.

```suggestion
this approach will make it impossible to use the official ROS Ubuntu packages that depend on Gazebo.
```

Reviewer 2: Another possible suggestion:
```suggestion
this approach will make it impossible to use some of the official ROS Ubuntu packages.
```

We do not recommend this approach for beginners!
</div>

To select a different release of Gazebo than the one officially supported by
your ROS distribution you must **manually compile**
[`ros_gz`](https://github.com/gazebosim/ros_gz) from source. This rule applies for
every  ROS packages that uses a Gazebo library. This approach may also require
that you modify your  ROS or Gazebo source code to support this compilation.

### Gazebo Garden (Not Recommended)

Gazebo Garden can be used with ROS 2 Humble and ROS 2 Rolling, but
[`ros_gz`](https://github.com/gazebosim/ros_gz) will need to be compiled
from source.

 * Folow [these instruction to install gz-garden](https://gazebosim.org/docs/garden/install_ubuntu#binary-installation-on-ubuntu)
   from osrfoundation.org repository
 * Install [rosdep rules for Gazebo Garden](https://github.com/osrf/osrf-rosdep#installing-rosdep-rules-to-resolve-gazebo-garden-libraries)
 * Follow the instructions to compile `ros_gz` from source in a colcon workspace
   * [ROS 2 Humble](https://github.com/gazebosim/ros_gz/tree/humble#from-source)

>     * Be sure of using `export GZ_VERSION=garden`

Reviewer 2: tabulation seems broken

```suggestion
     * Be sure of using `export GZ_VERSION=garden`
```

   * Gazebo Garden

This means that including the osrfoundation repository is not strictly needed
to get the Gazebo binary packages, as it can be installed from the ROS
repository.

## Installing the default Gazebo version for a ROS distribution using binary installations

For users that need to run a specific version of ROS and want to use all

> the gazebo ROS related packages out-of-the-box, this is the recommended taking
> the following approach

Reviewer: I think this needs to be reworded. Maybe
```suggestion
the gazebo ROS related packages out-of-the-box, taking
the following approach is recommended
```

of Gazebo for ROS 2 Foxy is Citadel.To install Gazebo Citadel with Foxy simply
install `ros-foxy-ros-gz` from `packages.ros.org`. This package will
automatically install both Gazebo Citadel and the ros-gz bridge.

### ROS 1 Noetic
ROS Noetic officially supports Gazebo Citadel, meaning the recommended version
of Gazebo for ROS Noetic is Citadel.To install Gazebo Citadel with Noetic simply
install `ros-noetic-ros-gz` from `packages.ros.org`. This package will
automatically install both Gazebo Citadel and the ros-gz bridge.

## Using a specific and unsupported Gazebo version with ROS 2
<div class="warning">
<strong>Warning:</strong> Only use this approach if you absolutely need to run
a version of Gazebo that is not officially supported by your ROS distro. Using
this approach will make it impossible to use the official ROS Ubuntu packages that depend on Gazebo.

We do not recommend this approach for beginners!
</div>

To select a different release of Gazebo than the one officially supported by
your ROS distribution you must **manually compile**
[`ros_gz`](https://github.com/gazebosim/ros_gz) from source. This rule applies for

> every  ROS packages that uses a Gazebo library. This approach may also require

Reviewer: ```suggestion
every  ROS package that uses a Gazebo library. This approach may also require
```
